cluster/gce/util.sh

```diff
@@ -3082,13 +3162,212 @@ function create-master() {
+  mkdir -p ${CERT_TEMP}
+  create-certs "${TPSERVER_RESERVED_IP[$tp_sequence]}" "${TPSERVER_RESERVED_INTERNAL_IP[$tp_sequence]}"
+  create-etcd-certs "${TPSERVER_NAME[$tp_sequence]}" "" "" "${TPSERVER_RESERVED_INTERNAL_IP[$tp_sequence]}"
+  create-etcd-apiserver-certs "etcd-${TPSERVER_NAME[$tp_sequence]}" "${TPSERVER_NAME[$tp_sequence]}" "" "" "${TPSERVER_RESERVED_INTERNAL_IP[$tp_sequence]}"
+  create-scaleoutserver-instance "${TPSERVER_NAME[$tp_sequence]:-}" "${TPSERVER_RESERVED_IP[$tp_sequence]:-}" "${TPSERVER_RESERVED_INTERNAL_IP[$tp_sequence]:-}"
+
+}
+
+function create-rpmaster {
+  local rp_sequence=${1:-1}
+  NODE_TAG="${CLUSTER_NAME}-minion"
+  INITIAL_ETCD_CLUSTER="${MASTER_NAME}"
+  KUBERNETES_MASTER_NAME="${RPSERVER_RESERVED_IP[${rp_sequence}]}"
+  MASTER_ADVERTISE_ADDRESS="${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]}"
+  KUBERNETES_MASTER_INTERNAL_IP="${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]}"
+  KUBECONFIG="${RESOURCE_DIRECTORY}/kubeconfig${KUBEMARK_PREFIX}.rp-$rp_sequence"
+
+  if [[ "${REGISTER_MASTER_KUBELET:-}" == "true" ]]; then
+    KUBELET_APISERVER="${RPSERVER_RESERVED_IP[${rp_sequence}]}"
+  fi
+
+  MASTER_EXTRA_METADATA=${KUBE_MASTER_EXTRA_METADATA:-${KUBE_EXTRA_METADATA:-}}
+  if [[ "${MASTER_EXTRA_METADATA}" != "" ]]; then
+    MASTER_EXTRA_METADATA="${MASTER_EXTRA_METADATA},${TPCONFIG_METADATA}"
+  else
+    MASTER_EXTRA_METADATA="${TPCONFIG_METADATA}"
+  fi
+
+  CERT_TEMP="${KUBE_TEMP}/rp$rp_sequence"
+  mkdir -p ${CERT_TEMP}
+  create-certs "${RPSERVER_RESERVED_IP[$rp_sequence]}" "${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]}"
+  create-etcd-certs "${RPSERVER_NAME[$rp_sequence]}" "" "" "${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]}"
+  create-etcd-apiserver-certs "etcd-${RPSERVER_NAME[$rp_sequence]}" "${RPSERVER_NAME[$rp_sequence]}" "" "" "${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]}"
+  create-scaleoutserver-instance "${RPSERVER_NAME[$rp_sequence]:-}" "${RPSERVER_RESERVED_IP[$rp_sequence]:-}" "${RPSERVER_RESERVED_INTERNAL_IP[$rp_sequence]:-}"
+
+}
+
+# master machine name format: {KUBE_GCE_ZONE}-kubemark-tp-1-master
+# destination file " --resource-providers=/etc/srv/kubernetes/kube-scheduler/rp-kubeconfig"
+# TODO: avoid calling GCE compute from here
+# TODO: currently the same kubeconfig is used by both scheduler and kube-controller-managers on RP clusters
+#       Pending design on how RP kubeconfigs to be used on the TP cluster:
+#       if the current approach continue to be used,  modify the kubeconfigs so the scheduler and controllers
+#       can have different identity for refined RBAC and logging purposes
+#       if future design is to let scheduler and controller managers to point to a generic server to get those RP
+#       kubeconfigs, the generic service should generate different ones for them.
+# TODO: remove this restart because all tp/rp ip has been reserved and this config should be ready before tp/rp started
+function restart_tp_scheduler_and_controller {
```

Comment:
do we still need it with the prepare master where the master IPs are all set ?

Reply:
will re-org this section in follow-up optimized PR